src/trio/_tests/test_exports.py

```diff
@@ -572,3 +573,77 @@ def test_classes_are_final() -> None:
                 continue
 
             assert class_is_final(class_)
+
+
+def test_pyright_recognizes_init_attributes() -> None:
+    # Obviously, this isn't completely accurate.
+    # It should still be good enough. Hopefully.
+    # (attrs updates fields before we can access them)
+    files = []
+
+    parent = (Path(inspect.getfile(trio)) / "..").resolve()
+    for path in parent.glob("**/*.py"):
+        if "_tests" in str(path)[len(str(parent)) :]:
+            continue
+
+        with open(path, "rb") as f:
+            files.append(list(tokenize.tokenize(f.readline)))
+
+    for module in PUBLIC_MODULES:
+        for name, class_ in module.__dict__.items():
+            if not attrs.has(class_):
+                continue
+            if isinstance(class_, _util.NoPublicConstructor):
+                continue
+
+            file = None
+            start = None
+            for contents in files:
+                last_was_class = False
+                for i, token in enumerate(contents):
+                    if (
+                        token.type == tokenize.NAME
+                        and token.string == name
+                        and last_was_class
+                    ):
+                        assert file is None
+                        file = contents
+                        start = i - 1
+
+                    if token.type == tokenize.NAME and token.string == "class":
+                        last_was_class = True
+                    else:
+                        last_was_class = False
+
+            assert file is not None, f"{name}: {class_!r}"
+            assert start is not None
+
+            count = -1
+            # linters don't like my not using the index, go figure.
+            for end_offset, token in enumerate(file[start:]):  # noqa: B007
```

Comment:
if you declare `end_offset` outside the loop I think the warning goes away (and makes the code easier to read).

codecov wants a `pragma: no branch` on the loop to hit 100%

Reply:
Unfortunately the warning doesn't go away; it does make the code more correct (previously it didn't handle `start==len(file)-1`, I think)

Reply:
I think with the zero-initialization now that should be good

Reply:
This is fairly gnarly to fix in an AST visitor, I don't see any way of doing it that's not overkill, but I think you could open an issue for ruff and see if they have a decent way of getting at it.